src/Language/PureScript/Terms.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- Data types for PureScript terms
+--
+module Language.PureScript.Terms (PSString, mkString, renderJSON, codePoints, toLabel, toUTF16CodeUnits) where
+
+import qualified Data.Aeson as A
+import Prelude.Compat
+import Data.Monoid ((<>))
+import Data.String (IsString(..))
+import Data.Word
+import Data.Text (Text)
+import qualified Data.Text as T
+import Numeric (showHex)
+import Language.PureScript.Types (Label(..))
+
+-- |
+-- Strings in PureScript are sequences of UTF-16 code units, which do not
+-- necessarily represent UTF-16 encoded text. For example, it is permissible
+-- for a string to contain *lone surrogates,* i.e. characters in the range
+-- 0xD800 - 0xDFFF which do not appear as a part of a surrogate pair.
+newtype PSString = PSString [Word16]
+  deriving (Eq, Ord, A.FromJSON)
+
+instance Show PSString where
+  show = T.unpack . renderJSON
+
+instance IsString PSString where
+  fromString a = PSString $ concatMap encodeUTF16 a
+    where
+    surrogates :: Char -> (Word16, Word16)
+    surrogates c = (toWord (h + 0xD800), toWord (l + 0xDC00))
+      where (h, l) = divMod (fromEnum c - 0x10000) 0x400
+
+    encodeUTF16 :: Char -> [Word16]
+    encodeUTF16 c | fromEnum c > 0xFFFF = [high, low]
+      where (high, low) = surrogates c
+    encodeUTF16 c = [toWord $ fromEnum c]
+
+instance A.ToJSON PSString where
+  toJSON = A.toJSON . map toChar . toUTF16CodeUnits
+
+renderJSON :: PSString -> Text
+renderJSON s = T.pack $ "\"" <> concatMap encodeChar (toUTF16CodeUnits s) <> "\""
+  where
+  encodeChar :: Word16 -> String
+  encodeChar c | c > 0xFFF = "\\u" <> showHex c ""
+  encodeChar c | c > 0xFF = "\\u0" <> showHex c ""
+  encodeChar c | c < 0x10 = "\\x0" <> showHex c ""
+  encodeChar c | c > 0x7E || c < 0x20 = "\\x" <> showHex c ""
+  encodeChar c | toChar c == '\b' = "\\b"
+  encodeChar c | toChar c == '\t' = "\\t"
+  encodeChar c | toChar c == '\n' = "\\n"
+  encodeChar c | toChar c == '\v' = "\\v"
+  encodeChar c | toChar c == '\f' = "\\f"
+  encodeChar c | toChar c == '\r' = "\\r"
+  encodeChar c | toChar c == '"'  = "\\\""
+  encodeChar c | toChar c == '\\' = "\\\\"
+  encodeChar c = pure $ toChar c
+
+toChar :: Word16 -> Char
+toChar = toEnum . fromIntegral
+
+toWord :: Int -> Word16
+toWord = fromIntegral
+
+toInt :: Word16 -> Int
+toInt = fromIntegral
+
+mkString :: Text -> PSString
+mkString = fromString . T.unpack
+
+-- TODO: remove me, replace with toLabel + possibly others
+codePoints :: PSString -> Text
```

Comment:
Does it make sense to have this return a `Maybe Text`, since it attempts to convert a `PSString` into a `Text`, but not all `PSString` values can be represented as `Text` (specifically, PSStrings which contain lone surrogates)?